python/ray/data/_internal/arrow_block.py

```diff
@@ -207,7 +231,23 @@ def numpy_to_block(
             col = convert_udf_returns_to_numpy(col)
             # Use Arrow's native *List types for 1-dimensional ndarrays.
             if col.dtype.type is np.object_ or col.ndim > 1:
-                col = ArrowTensorArray.from_numpy(col, col_name)
+                try:
+                    col = ArrowTensorArray.from_numpy(col, col_name)
+                except (
+                    pyarrow.ArrowInvalid,
+                    pyarrow.ArrowNotImplementedError,
+                    pyarrow.ArrowTypeError,
+                ):
+                    if object_extension_type_allowed():
+                        if log_once(f"arrow_object_pickle_{col_name}"):
+                            logger.warning(
+                                f"Failed to interpret {col_name} as "
+                                "multi-dimensional arrays. It will be pickled."
```

Comment:
The current code looks at any column which has a numpy dtype of object and checks if it can be converted into a list of tensors. If it can't be, then we fall back to ArrowPythonObjectArray. This happens when the numpy array has a dtype of object because you have custom objects in there.